bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BUnionType.java

```diff
@@ -108,11 +131,22 @@ public boolean equals(Object o) {
         if (this == o) {
             return true;
         }
+
         if (!(o instanceof BUnionType)) {
             return false;
         }
+
         BUnionType that = (BUnionType) o;
-        return memberTypes.containsAll(that.memberTypes) && that.memberTypes.containsAll(memberTypes);
+        if (this.memberTypes.size() != that.memberTypes.size()) {
+            return false;
+        }
+
+        for (int i = 0; i < memberTypes.size(); i++) {
+            if (!this.memberTypes.get(i).equals(that.memberTypes.get(i))) {
```

Comment:
Here we check for whether the two unions are the 'same' rather than the assignability/equality. So the simpler case would be fast. If it fails, then it will always fall back to the assignable check. That would satisfy the equality of  `int|string` and `string|int`